jetbrains-ultimate/src/software/aws/toolkits/jetbrains/services/lambda/nodejs/NodeJsHelper.kt

```diff
@@ -0,0 +1,29 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.services.lambda.nodejs
+
+import com.intellij.lang.javascript.buildTools.npm.PackageJsonUtil
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.roots.ProjectFileIndex
+import com.intellij.openapi.vfs.VirtualFile
+
+/**
+ * WebStorm doesn't allow user to mark a folder as source root. This method infers a folder as source root based on
+ * whether it has package.json file in it. If there is no package.json found in the path, content root will be returned.
+ *
+ * @param project The Node.js project.
+ * @param virtualFile The Node.js source code file.
+ * @return The inferred source root that contains package.json file, or content root of the file.
+ * @throws IllegalStateException If the contentRoot cannot be located
+ */
+fun inferSourceRoot(project: Project, virtualFile: VirtualFile): VirtualFile {
+    val projectFileIndex = ProjectFileIndex.getInstance(project)
+    val contentRoot = projectFileIndex.getContentRootForFile(virtualFile) ?: throw IllegalStateException("Cannot locate content root for file")
+
+    var parentFolder = virtualFile.parent
+    while (PackageJsonUtil.findChildPackageJsonFile(parentFolder) == null && parentFolder != contentRoot) {
```

Comment:
This assignment / mutation is a bit nasty - maybe recursion?

```kt
fun findPackageJson(file: VirtualFile): VirtualFile? {
    return when {
        PackageJsonUtil.findChildPackageJsonFile(file) != null -> file
        file == contentRoot -> file
        else -> findPackageJson(file.parent)
    }
}
return findPackageJson(virtualFile.parent)
```

Reply:
Not a big fun of recursion, but yeah it's good to keep all variables immutable. Updated.